In [1]:
!pip install bert_score

  Using cached nvidia_cuda_nvrtc_cu12-12.1.105-py3-none-manylinux1_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_cuda_runtime_cu12-12.1.105-py3-none-manylinux1_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_cuda_cupti_cu12-12.1.105-py3-none-manylinux1_x86_64.whl.metadata (1.6 kB)
  Using cached nvidia_cudnn_cu12-8.9.2.26-py3-none-manylinux1_x86_64.whl.metadata (1.6 kB)
  Using cached nvidia_cublas_cu12-12.1.3.1-py3-none-manylinux1_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_cufft_cu12-11.0.2.54-py3-none-manylinux1_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_curand_cu12-10.3.2.106-py3-none-manylinux1_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_cusolver_cu12-11.4.5.107-py3-none-manylinux1_x86_64.whl.metadata (1.6 kB)
  Using cached nvidia_cusparse_cu12-12.1.0.106-py3-none-manylinux1_x86_64.whl.metadata (1.6 kB)
  Using cached nvidia_nccl_cu12-2.20.5-py3-none-manylinux2014_x86_64.whl.metadata (1.8 kB)
  Using cached nvidia_nvtx_cu12-12.1.105-py3-none-manylinu

In [2]:
import bert_score

# Example candidate responses from your chatbot
candidates = [
    "To update your Ubuntu system, you can follow these steps: 1. Open the Terminal by pressing `Ctrl + Alt + T` on your keyboard. 2. Type the following command to update the package lists for upgrades and new package installations: sudo apt update. 3. After the update is complete, you can upgrade your system packages to the latest versions by running: sudo apt upgrade. 4. If there are kernel updates available, you should also run the following command to upgrade your kernel: sudo apt dist-upgrade. 5. Finally, to ensure your system is fully up to date, you should also run: sudo apt full-upgrade. 6. Reboot your system if necessary to apply any kernel updates. These steps will help you keep your Ubuntu system updated with the latest software and security patches.",
    "To install software in Ubuntu, you can use the Software Center or the terminal. Here are the steps for both methods: 1. Using Software Center: - Click on the 'Ubuntu Software' icon in the launcher. - Search for the software you want to install. - Click on the software you want to install and then click the 'Install' button. - Enter your password if prompted. - The software will be downloaded and installed on your system. 2. Using the Terminal: - Open the terminal by pressing Ctrl + Alt + T. - Use the following command to update the package list: sudo apt update. - Use the following command to install a software package. Replace 'package-name' with the name of the software you want to install: sudo apt install package-name. - Enter your password if prompted and then confirm the installation. After following these steps, the software should be successfully installed on your system.",
    "One of the main differences between Ubuntu and other Linux distributions is the user interface and default software packages they come with. Ubuntu, for example, uses the GNOME desktop environment by default, while other distributions may use different desktop environments like KDE, XFCE, or Cinnamon. Additionally, Ubuntu has its own set of pre-installed software and package management tools that differentiate it from other distributions.",
    "To set up a static IP address on Ubuntu, you can follow these steps: 1. Open the terminal by pressing Ctrl + Alt + T. 2. Edit the network configuration file by running the command: sudo nano /etc/netplan/01-network-manager-all.yaml. 3. Within the file, you should see something similar to the following: network: version: 2 renderer: NetworkManager ethernets: enp0s3: dhcp4: yes. 4. Change the configuration to specify a static IP address. For example: network: version: 2 renderer: NetworkManager ethernets: enp0s3: dhcp4: no addresses: [192.168.1.100/24] gateway4: 192.168.1.1 nameservers: addresses: [8.",
    "To check disk usage on Ubuntu, you can use the df command in the terminal. Here's how you can do it: 1. Open a terminal window by pressing `Ctrl + Alt + T` or by searching for 'Terminal' in the applications menu. 2. In the terminal, type the following command and press Enter: df -h. This command will display the disk usage information in a human-readable format, showing you details such as total disk space, used space, available space, and usage percentages for each mounted filesystem. You can also use the du command to check disk usage of specific directories. For example, to check the disk usage of the current directory, you can use: du -h. These commands will help you monitor your disk usage and manage your storage more effectively.",
    "To install a new package on Ubuntu, you can use the following steps: 1. Open a terminal window by pressing Ctrl+Alt+T. 2. Update the package list by running the command: sudo apt update. 3. Install the package by running: sudo apt install <package-name>. Replace <package-name> with the name of the package you want to install. 4. Enter your password when prompted and press Enter. 5. Confirm the installation by typing y and pressing Enter. After following these steps, the package should be installed on your Ubuntu system."
]

# Corresponding reference responses (ideal answers)
references = [
    "To update your Ubuntu system, open the terminal by pressing `Ctrl + Alt + T`. First, update the package lists with the command `sudo apt update`. Then, upgrade the installed packages with `sudo apt upgrade`. If there are kernel or major upgrades, use `sudo apt dist-upgrade` to apply them. Finally, run `sudo apt full-upgrade` to ensure all packages are updated. Reboot your system if required.",
    "You can install software on Ubuntu using the Software Center or via the terminal. For the Software Center, open it, search for the desired software, and click 'Install.' Alternatively, in the terminal, first run `sudo apt update`, then install the package with `sudo apt install package-name`, replacing 'package-name' with the software you want to install.",
    "Ubuntu distinguishes itself from other Linux distributions by its use of the GNOME desktop environment, a user-friendly interface, and a large community support base. Additionally, Ubuntu emphasizes ease of use and regular updates, making it suitable for both beginners and advanced users. Other distributions may use different desktop environments and package managers, catering to different user needs and preferences.",
    "To set up a static IP address on Ubuntu, open the terminal and edit the Netplan configuration file with `sudo nano /etc/netplan/01-netcfg.yaml`. Modify the file to disable DHCP and specify the static IP, gateway, and DNS servers. For example: network: version: 2 renderer: NetworkManager ethernets: enp0s3: dhcp4: no addresses: [192.168.1.100/24] gateway4: 192.168.1.1 nameservers: addresses: [8.8.8.8, 8.8.4.4] After saving the file, apply the changes with `sudo netplan apply`.",
    "To check disk usage on Ubuntu, open a terminal and use the `df -h` command. This will display disk usage information in a human-readable format, showing total, used, and available space for each filesystem. For detailed usage of a specific directory, use `du -sh /path/to/directory` to see the size of the directory and its contents.",
    "To install a new package on Ubuntu, open the terminal and first update your package list with `sudo apt update`. Then, install the package with `sudo apt install package-name`, replacing 'package-name' with the name of the software. Confirm the installation when prompted, and the package will be installed on your system."
]

# Calculate BERTScore
P, R, F1 = bert_score.score(candidates, references, lang="en", rescale_with_baseline=True)

# Print the BERTScore for each response
for i, (p, r, f1) in enumerate(zip(P, R, F1)):
    print(f"Response {i+1}:")
    print(f"  Precision: {p.item():.4f}")
    print(f"  Recall: {r.item():.4f}")
    print(f"  F1 Score: {f1.item():.4f}\n")


/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/482 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.42G [00:00<?, ?B/s]

Some weights of RobertaModel were not initialized from the model checkpoint at roberta-large and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Response 1:
  Precision: 0.4528
  Recall: 0.6456
  F1 Score: 0.5482

Response 2:
  Precision: 0.2420
  Recall: 0.5028
  F1 Score: 0.3702

Response 3:
  Precision: 0.3824
  Recall: 0.4125
  F1 Score: 0.3984

Response 4:
  Precision: 0.4259
  Recall: 0.4635
  F1 Score: 0.4455

Response 5:
  Precision: 0.3523
  Recall: 0.5122
  F1 Score: 0.4320

Response 6:
  Precision: 0.4133
  Recall: 0.6797
  F1 Score: 0.5440



# **Summary of BERTScore Results**

Response 1:

Precision: 0.4528
Recall: 0.6456
F1 Score: 0.5482

Analysis: This response has a relatively high recall, indicating that it covers a large portion of the relevant information found in the reference response. The F1 Score is also the highest among all responses, showing a good balance between precision and recall.

Response 2:

Precision: 0.2420
Recall: 0.5028
F1 Score: 0.3702

Analysis: The lower precision and F1 Score suggest that this response includes a fair amount of information not found in the reference, or it may have less relevant or less accurate content. The recall is moderate, showing some coverage of the reference content.

Response 3:

Precision: 0.3824
Recall: 0.4125
F1 Score: 0.3984

Analysis: This response has relatively balanced precision and recall but is neither high. The response captures some of the relevant content but misses out on critical details, resulting in a moderate F1 Score.

Response 4:

Precision: 0.4259
Recall: 0.4635
F1 Score: 0.4455

Analysis: This response shows a slightly better precision than recall, with a moderate F1 Score. It suggests that the response is somewhat relevant but might lack completeness or depth compared to the reference.

Response 5:

Precision: 0.3523
Recall: 0.5122
F1 Score: 0.4320

Analysis: This response has a moderate recall and lower precision, indicating that it covers some relevant information but might also include irrelevant content. The F1 Score reflects a moderate performance.

**Overall Evaluation:**

Response 1 stands out as the strongest in terms of both recall and F1 Score, indicating that it is the most aligned with the reference.
Response 2 has the lowest precision and F1 Score, suggesting that it is less accurate and relevant compared to the reference.
Response 3, 4, and 5 show moderate alignment with the reference but could be improved in terms of either precision, recall, or both.

In [1]:
!pip install rouge-score


  Preparing metadata (setup.py) ... done
  Created wheel for rouge-score: filename=rouge_score-0.1.2-py3-none-any.whl size=24935 sha256=98b5c9b3d018652a1b2a31d9b792a7a4a121a3560ac0ffd915bf4bb33975270a
  Stored in directory: /root/.cache/pip/wheels/5f/dd/89/461065a73be61a532ff8599a28e9beef17985c9e9c31e541b4
Successfully built rouge-score


In [2]:
from rouge_score import rouge_scorer

# Example candidate responses from your chatbot
candidates = [
    "To update your Ubuntu system, you can follow these steps: 1. Open the Terminal by pressing `Ctrl + Alt + T` on your keyboard. 2. Type the following command to update the package lists for upgrades and new package installations: sudo apt update. 3. After the update is complete, you can upgrade your system packages to the latest versions by running: sudo apt upgrade. 4. If there are kernel updates available, you should also run the following command to upgrade your kernel: sudo apt dist-upgrade. 5. Finally, to ensure your system is fully up to date, you should also run: sudo apt full-upgrade. 6. Reboot your system if necessary to apply any kernel updates. These steps will help you keep your Ubuntu system updated with the latest software and security patches.",
    "To install software in Ubuntu, you can use the Software Center or the terminal. Here are the steps for both methods: 1. Using Software Center: - Click on the 'Ubuntu Software' icon in the launcher. - Search for the software you want to install. - Click on the software you want to install and then click the 'Install' button. - Enter your password if prompted. - The software will be downloaded and installed on your system. 2. Using the Terminal: - Open the terminal by pressing Ctrl + Alt + T. - Use the following command to update the package list: sudo apt update. - Use the following command to install a software package. Replace 'package-name' with the name of the software you want to install: sudo apt install package-name. - Enter your password if prompted and then confirm the installation. After following these steps, the software should be successfully installed on your system.",
    "One of the main differences between Ubuntu and other Linux distributions is the user interface and default software packages they come with. Ubuntu, for example, uses the GNOME desktop environment by default, while other distributions may use different desktop environments like KDE, XFCE, or Cinnamon. Additionally, Ubuntu has its own set of pre-installed software and package management tools that differentiate it from other distributions.",
    "To set up a static IP address on Ubuntu, you can follow these steps: 1. Open the terminal by pressing Ctrl + Alt + T. 2. Edit the network configuration file by running the command: sudo nano /etc/netplan/01-network-manager-all.yaml. 3. Within the file, you should see something similar to the following: network: version: 2 renderer: NetworkManager ethernets: enp0s3: dhcp4: yes. 4. Change the configuration to specify a static IP address. For example: network: version: 2 renderer: NetworkManager ethernets: enp0s3: dhcp4: no addresses: [192.168.1.100/24] gateway4: 192.168.1.1 nameservers: addresses: [8.",
    "To check disk usage on Ubuntu, you can use the df command in the terminal. Here's how you can do it: 1. Open a terminal window by pressing `Ctrl + Alt + T` or by searching for 'Terminal' in the applications menu. 2. In the terminal, type the following command and press Enter: df -h. This command will display the disk usage information in a human-readable format, showing you details such as total disk space, used space, available space, and usage percentages for each mounted filesystem. You can also use the du command to check disk usage of specific directories. For example, to check the disk usage of the current directory, you can use: du -h. These commands will help you monitor your disk usage and manage your storage more effectively.",
    "To install a new package on Ubuntu, you can use the following steps: 1. Open a terminal window by pressing Ctrl+Alt+T. 2. Update the package list by running the command: sudo apt update. 3. Install the package by running: sudo apt install <package-name>. Replace <package-name> with the name of the package you want to install. 4. Enter your password when prompted and press Enter. 5. Confirm the installation by typing y and pressing Enter. After following these steps, the package should be installed on your Ubuntu system."
]

# Corresponding reference responses (ideal answers)
references = [
    "To update your Ubuntu system, open the terminal by pressing `Ctrl + Alt + T`. First, update the package lists with the command `sudo apt update`. Then, upgrade the installed packages with `sudo apt upgrade`. If there are kernel or major upgrades, use `sudo apt dist-upgrade` to apply them. Finally, run `sudo apt full-upgrade` to ensure all packages are updated. Reboot your system if required.",
    "You can install software on Ubuntu using the Software Center or via the terminal. For the Software Center, open it, search for the desired software, and click 'Install.' Alternatively, in the terminal, first run `sudo apt update`, then install the package with `sudo apt install package-name`, replacing 'package-name' with the software you want to install.",
    "Ubuntu distinguishes itself from other Linux distributions by its use of the GNOME desktop environment, a user-friendly interface, and a large community support base. Additionally, Ubuntu emphasizes ease of use and regular updates, making it suitable for both beginners and advanced users. Other distributions may use different desktop environments and package managers, catering to different user needs and preferences.",
    "To set up a static IP address on Ubuntu, open the terminal and edit the Netplan configuration file with `sudo nano /etc/netplan/01-netcfg.yaml`. Modify the file to disable DHCP and specify the static IP, gateway, and DNS servers. For example: network: version: 2 renderer: NetworkManager ethernets: enp0s3: dhcp4: no addresses: [192.168.1.100/24] gateway4: 192.168.1.1 nameservers: addresses: [8.8.8.8, 8.8.4.4] After saving the file, apply the changes with `sudo netplan apply`.",
    "To check disk usage on Ubuntu, open a terminal and use the `df -h` command. This will display disk usage information in a human-readable format, showing total, used, and available space for each filesystem. For detailed usage of a specific directory, use `du -sh /path/to/directory` to see the size of the directory and its contents.",
    "To install a new package on Ubuntu, open the terminal and first update your package list with `sudo apt update`. Then, install the package with `sudo apt install package-name`, replacing 'package-name' with the name of the software. Confirm the installation when prompted, and the package will be installed on your system."
]

# Initialize the ROUGE scorer
scorer = rouge_scorer.RougeScorer(['rougeL'], use_stemmer=True)

# Calculate ROUGE-L for each candidate-reference pair
for i, (candidate, reference) in enumerate(zip(candidates, references)):
    score = scorer.score(reference, candidate)
    print(f"Response {i+1} ROUGE-L: {score['rougeL'].fmeasure:.4f}")


Response 1 ROUGE-L: 0.4490
Response 2 ROUGE-L: 0.3861
Response 3 ROUGE-L: 0.3200
Response 4 ROUGE-L: 0.5824
Response 5 ROUGE-L: 0.4339
Response 6 ROUGE-L: 0.5833


ROUGE-L Score Summary
Response 1:

ROUGE-L: 0.4490
Analysis: This response has a moderate overlap with the reference, capturing nearly half of the important sequences.
Response 2:

ROUGE-L: 0.3861
Analysis: This response has a lower overlap compared to Response 1, indicating that it misses or poorly aligns with the reference content.
Response 3:

ROUGE-L: 0.3200
Analysis: This response has the lowest ROUGE-L score, suggesting it has the least alignment with the reference in terms of important sequences.
Response 4:

ROUGE-L: 0.5824
Analysis: This response has a relatively high score, indicating a strong alignment with the reference. It effectively captures most of the important sequences.
Response 5:

ROUGE-L: 0.4339
Analysis: This response has a moderate overlap, similar to Response 1, but slightly lower.
Response 6:

ROUGE-L: 0.5833
Analysis: This response has the highest ROUGE-L score, showing the best alignment with the reference. It effectively captures the key sequences in the reference response.
Overall Evaluation:
Response 6 and Response 4 are the strongest, with the highest ROUGE-L scores, indicating they closely match the reference answers in terms of important sequences.
Response 3 has the lowest score, indicating it may need the most improvement to better align with the reference.
Responses 1 and 5 have moderate scores, suggesting they capture a decent amount of the relevant content but could still be refined for better alignment.